rfcs/0056-FE-APIs.md

```diff
@@ -0,0 +1,234 @@
+  funcName<TConfig extends object>(configFn: ConfigFn<TConfig>): void;
+}
+
+type ConfigFn<TConfig extends object> = (
+  prevConfigs: TConfig[],
+  ctx: EditViewContext
+) => TConfig[];
+```
+
+Because the function receives the previous list of configs, we expect the complete list of configs to be returned, this means if a plugin required it could remove a configuration, or mutate a current configuration building upon one another e.g. the publish action would be able to be mutated by the `i18n` plugin correctly.
+
+## **Injecting into the Header**
+
+A `HeaderAction` is a constrained API enabling users to inject a specific set of items into the header toolbar section e.g. a `Select` or `Button`. It’s specifically restricted and we can make it looser as requirements surface and are validated.
+
+```tsx
+interface StrapiApp {
+  addDocumentHeaderAction(action: HeaderAction[]): void;
+  addDocumentHeaderAction(actionFn: HeaderActionFn): void;
+}
+
+type HeaderActionFn = (
+  previousPanel: HeaderAction[],
+  ctx: EditViewContext
+) => HeaderAction[];
+
+type HeaderAction = ButtonAction | SelectAction;
+
+interface ButtonAction {
+  label: string;
+  icon?: React.ReactNode;
+  type: "icon" | "default";
+  onClick?: (
+    event: React.MouseEvent<HTMLButtonElement>,
+    ctx: EditViewContext
+  ) => void;
+}
+
+interface SelectAction {
+  label: string;
+  options: Array<{ value: string; label: string; textValue?: string }>;
+  onSelect?: (value: string, ctx: EditViewContext) => void;
+}
+```
+
+## Injecting into the SideBar
+
+A `PanelConfig` allows users to define a new configuration panel in the sidebar where they can put whatever they desire. e.g. Review Workflows could be it’s own panel.
+
+```tsx
+interface StrapiApp {
+  addEditViewSidePanel(panel: PanelConfig[]): void;
+  addEditViewSidePanel(panelFn: PanelConfigFn): void;
+}
+
+type PanelConfigFn = (
+  previousPanel: PanelConfig[],
+  ctx: EditViewContext
+) => PanelConfig[];
+
+interface PanelConfig {
+  title: string;
+  content: React.ReactNode;
+}
+```
+
+For backwards compatibility we’ll render all components still using the injection-zone API in the default panel that holds the Document Actions.
+
+## Extending Document Actions
+
+A `DocumentAction` is a configuration object we pass and is rendered as part of the ActionButton menu in the EditView.
+
+```tsx
+// Partial type to illustrate an overload.
+interface StrapiApp {
+  addDocumentAction(action: DocumentAction[]): void;
+  addDocumentAction(actionFunction: DocumentActionFunction): void;
+}
+
+type DocumentActionFunction = (
+  previousActions: DocumentAction[],
+  ctx: EditViewContext
+) => DocumentAction[];
+
+interface DocumentAction {
+  label: string;
+  onClick: (
+    event: React.MouseEvent<HTMLButtonElement>,
+    ctx: EditViewContext
+  ) => void;
+  icon?: React.ElementType;
+  disabled?: boolean;
+  dialog?: DialogOptions | NotifiationOptions | ModalOptions;
```

Comment:
there's definitely use-cases e.g. your document action is to sync with an external data source, you just want a notificaiton that says "synced"

Reply:
@joshuaellis it would always be synced or error. if it says synced while it has an error that would not be a good user experience.

or can dialog also be a function?